### PR TITLE
[Security] Refactor logout listener to dispatch an event instead

### LINKED ISSUE #22473: Inconsistency in security factories
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | all

I've implemented my own security authentication using a custom factory as e.g. described in the [docs here](http://symfony.com/doc/current/security/custom_authentication_provider.html).
I read the source code in `SecurityExtension` carefully and as far as I could see, thanks to the support for factories and the `AbstractFactory` I was able to quickly have my own listener, provider etc. registered. Configuring it in my `security.yml` was as easy as this:

```yml
firewall_name:
    pattern: ~
    anonymous: true
    my_factory_key: { my_factory_settings: foobar }
```

However, this does not work for `logout`. As you can see [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L351), `logout` has to be explicitly defined on the firewall so I cannot just use my factory to provide my own logout listeners. I would like to replace the logout listener because I don't want it to check for a path but for request parameters to be present.
I just wonder if this is on purpose or if this is actually an inconsistency? Maybe anyone can tell me :-)

### LINKED ISSUE #25212: [Security] Custom Logout listeners
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | 4.1

I've been implementing SSO logout in a public bundle, and it's struck me that there's no really good way for a public bundle to provide logout customizations.

In this bundle ([HappyrAuth0Bundle](https://github.com/Happyr/auth0-bundle)), after logging out in Symfony, I'd like to pass the user on to an endpoint on the SSO server to be logged out there, and pass the final target on to them, so the user is redirected first to the SSO server, then back to my Symfony application. 

If I was just writing a regular Symfony application, this would be trivially solvable with either a logout handler, or a logout success handler, but in a public bundle, I don't want to burden my users by having them configure or possibly write those. Ideally (and what I ended up implementing in [the changeset](https://github.com/rdey/auth0-bundle/commit/abc5d472f6be7395cf44ade11227e38d2282067a)) I'd want my users to be able to just write something like the following in their `security.yml`.

```
security:
    firewalls:
        secured_area:
            pattern: ^/
            auth0_sso:
                check_path: login_check
            auth0_logout:
                path: logout
                target: /
```

This is somewhat doable (see the changeset above), but in order to do so, I had to create a fake provider, use an interface for something it clearly wasn't meant to (SecurityFactoryInterface), and even then, there's no way to fully replicate the behavior as in `logout`, because I cannot get a hold of which firewall context is currently used when creating the services.

There's also the possibility of wanting to disable functionality in the logout listener. In my `auth0_logout` section I don't allow for a logout success handler, because it doesn't make sense in an SSO context to allow a custom response; the response always has to be a redirect to the SSO service.

## Discussion points
* Am I wrong in thinking that it's necessary for good DX to provide a special `auth0_logout` key in the firewall configuration?
* If not, what is the preferable approach in making Symfony allow bundles to provide custom logouters? Should I focus on making it possible for SecurityFactories to not return providers and entrypoints, or should I create a new interface specifically for allowing custom logout providers?

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes (sort of...)
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Fix #25212, Fix #22473
| License       | MIT
| Doc PR        | tbd

The current `LogoutListener` has some extension points, but they are not really DX-friendly (ref #25212). It requires hacking a `addMethodCall('addHandler')` in the container builder to register a custom logout handler.  
Also, it is impossible to overwrite the default logout functionality from a bundle (ref #22473).

This PR introduces a `LogoutEvent` that replaces both the `LogoutSuccessHandlerInterface` and `LogoutHandlerInterface`. This provides a DX-friendly extension point and also cleans up the authentication factories (no more `addMethodCall()`'s).

In order to allow different logout handlers for different firewalls, I created a specific event dispatcher for each firewall (as also shortly discussed in #33558). The `dispatcher` tag attribute allows you to specify which dispatcher it should be registered to (defaulting to the global dispatcher). The `EventBubblingLogoutListener` also dispatches logout events on the global dispatcher, to be used for listeners that should run on all firewalls.

_@weaverryan and I discussed this feature while working on #33558, but figured it was unrelated and could be done while preservering BC. So that's why a separate PR is created._

### CODE REVIEW

**Comment by wouterj** [comment]
_the fabbot exception message error is a false-positive_

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 1]
```diff
@@ -34,31 +37,47 @@ class LogoutListener extends AbstractListener
...
-        $this->successHandler = $successHandler;
         $this->csrfTokenManager = $csrfTokenManager;
-        $this->handlers = [];
+        $this->eventDispatcher = $eventDispatcher;
     }
 
     public function addHandler(LogoutHandlerInterface $handler)
     {
-        $this->handlers[] = $handler;
+        trigger_deprecation('symfony/security-http', '5.1', 'Calling "%s" is deprecated, register a listener on the "%s" event instead.', __METHOD__, LogoutEvent::class);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: is the methode deprecated also? if yes, we should add the annotation also

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 2]
```diff
@@ -78,9 +97,9 @@ public function supports(Request $request): ?bool
      * @throws LogoutException   if the CSRF token is invalid
      * @throws \RuntimeException if the LogoutSuccessHandlerInterface instance does not return a response
      */
-    public function authenticate(RequestEvent $event)
+    public function authenticate(RequestEvent $requestEvent)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: can you please keep the original name, to reduce the diff?

File: UPGRADE-5.1.md [hunk 3]
```diff
@@ -91,6 +91,8 @@ Security
    {% endif %}
    ```
 
+ * Deprecated `LogoutSuccessHandlerInterface` and `LogoutHandlerInterface`. Register a listener on the `LogoutEvent` event instead.
```
Comment (nicolas-grekas) [inline UPGRADE-5.1.md]: UPGRADE-6.0 should be updated too

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php [hunk 4]
```diff
@@ -116,6 +110,11 @@ public function create(ContainerBuilder $container, string $id, array $config, ?
         $listener->replaceArgument(1, new Reference($rememberMeServicesId));
         $listener->replaceArgument(5, $config['catch_exceptions']);
 
+        // remember-me logout listener
+        $container->register(RememberMeLogoutListener::class)
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php]: In core, we must always give an id to service - we don't do FQCN service (if we did, we should fix it). This allows proper separation of concerns (DI referencing vs class naming)

Why doesn't the id vary by id btw? If it's normal,can't the service be declared in XML?

File: src/Symfony/Bundle/SecurityBundle/Security/LegacyLogoutHandlerListener.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+use Symfony\Component\Security\Http\Logout\LogoutHandlerInterface;
+use Symfony\Component\Security\Http\Logout\LogoutSuccessHandlerInterface;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @internal
+ *
+ * @deprecated to be removed in version 6.0
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/Security/LegacyLogoutHandlerListener.php]: Not sure it makes sense to deprecate and introduce in the same PR. If it's internal, this notice should be removed.

File: src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php [hunk 6]
```diff
@@ -114,12 +119,20 @@ public function process(ContainerBuilder $container)
             }
             $class = $r->name;
 
+            $dispatcherDefinition = $globalDispatcherDefinition;
+            foreach ($tags as $attributes) {
+                if (isset($attributes['dispatcher'])) {
+                    $dispatcherDefinition = $container->getDefinition($attributes['dispatcher']);
+                    break;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: Not sure about this logic in the foreach: this creates a correlation between tags, while they should be independent to me. Otherwise, we don't allow one listener to subscribe to several dispatchers, which looks like a serious limitation to me.

File: src/Symfony/Component/Security/CHANGELOG.md [hunk 7]
```diff
@@ -7,6 +7,8 @@ CHANGELOG
  * Added access decision strategy to override access decisions by voter service priority
  * Added `IS_ANONYMOUS`, `IS_REMEMBERED`, `IS_IMPERSONATOR`
  * Hash the persistent RememberMe token value in database.
+ * Added LogoutEvent to allow custom logout listeners.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/CHANGELOG.md]: missing backticks

File: src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public function onLogout(LogoutEvent $event): void
+    {
+        if (!$response = $event->getResponse()) {
+            return;
+        }
+
+        foreach ($this->cookies as $cookieName => $cookieData) {
+            $response->headers->clearCookie($cookieName, $cookieData['path'], $cookieData['domain']);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php]: we'd also need to pass the other attributes to properly clear any cookies, see #36173

File: src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+
+/**
+ * This listener clears the passed cookies when a user logs out.
+ *
+ * @author Johannes M. Schmitt <schmittjoh@gmail.com>
+ *
+ * @internal
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php]: is this required? `final` instead?

File: src/Symfony/Component/Security/Http/EventListener/CsrfTokenClearingLogoutListener.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Csrf\TokenStorage\ClearableTokenStorageInterface;
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+
+/**
+ * @author Christian Flothmann <christian.flothmann@sensiolabs.de>
+ *
+ * @internal
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/CsrfTokenClearingLogoutListener.php]: is this required? `final` instead?

File: src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * @param array $cookies An array of cookies (keys are names, values contain path and domain) to unset
+     */
+    public function __construct(array $cookies)
+    {
+        $this->cookies = $cookies;
+    }
+
+    public static function getSubscribedEvents(): array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php]: should be listed last to me

File: src/Symfony/Component/Security/Http/EventListener/CsrfTokenClearingLogoutListener.php [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+class CsrfTokenClearingLogoutListener implements EventSubscriberInterface
+{
+    private $csrfTokenStorage;
+
+    public function __construct(ClearableTokenStorageInterface $csrfTokenStorage)
+    {
+        $this->csrfTokenStorage = $csrfTokenStorage;
+    }
+
+    public static function getSubscribedEvents(): array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/CsrfTokenClearingLogoutListener.php]: should be listed last to me

File: src/Symfony/Component/Security/Http/EventListener/DefaultLogoutListener.php [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+use Symfony\Component\Security\Http\Event\LogoutEvent;
+use Symfony\Component\Security\Http\HttpUtils;
+
+/**
+ * Default logout listener will redirect users to a configured path.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Alexander <iam.asm89@gmail.com>
+ *
+ * @internal
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/DefaultLogoutListener.php]: same :)

File: src/Symfony/Component/Security/Http/EventListener/SessionLogoutListener.php [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Handler for clearing invalidating the current session.
+ *
+ * @author Johannes M. Schmitt <schmittjoh@gmail.com>
+ *
+ * @internal
+ */
+class SessionLogoutListener implements EventSubscriberInterface
+{
+    public static function getSubscribedEvents(): array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/EventListener/SessionLogoutListener.php]: should be listed last to me

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 15]
```diff
@@ -34,31 +37,47 @@ class LogoutListener extends AbstractListener
...
-    private $successHandler;
     private $httpUtils;
     private $csrfTokenManager;
+    private $eventDispatcher;
 
     /**
      * @param array $options An array of options to process a logout attempt
      */
-    public function __construct(TokenStorageInterface $tokenStorage, HttpUtils $httpUtils, LogoutSuccessHandlerInterface $successHandler, array $options = [], CsrfTokenManagerInterface $csrfTokenManager = null)
+    public function __construct(TokenStorageInterface $tokenStorage, HttpUtils $httpUtils, /* EventDispatcherInterface */$eventDispatcher, array $options = [], CsrfTokenManagerInterface $csrfTokenManager = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: let's add it to the docblock also

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 16]
```diff
@@ -34,31 +37,47 @@ class LogoutListener extends AbstractListener
...
     {
+        if (!$eventDispatcher instanceof EventDispatcherInterface) {
+            trigger_deprecation('symfony/security-http', '5.1', 'Passing a logout success handler to "%s" is deprecated, pass an instance of "%s" instead.', __METHOD__, EventDispatcherInterface::class);
+
+            if (!$eventDispatcher instanceof LogoutSuccessHandlerInterface) {
+                throw new \InvalidArgumentException(sprintf('Argument 3 of "%s" must be instance of "%s" or "%s", "%s" given.', __METHOD__, EventDispatcherInterface::class, LogoutSuccessHandlerInterface::class, get_debug_type($successHandler)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: `\TypeError`

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 17]
```diff
@@ -34,31 +37,47 @@ class LogoutListener extends AbstractListener
...
+            }
+
+            $successHandler = $eventDispatcher;
+            $eventDispatcher = new EventDispatcher();
+            $eventDispatcher->addListener(LogoutEvent::class, function (LogoutEvent $event) use ($successHandler) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: we still use named events, don't we?

Also, the closure can be made static.

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 18]
```diff
@@ -90,21 +109,17 @@ public function authenticate(RequestEvent $event)
             }
         }
 
-        $response = $this->successHandler->onLogoutSuccess($request);
-        if (!$response instanceof Response) {
-            throw new \RuntimeException('Logout Success Handler did not return a Response.');
-        }
+        $event = new LogoutEvent($request, $this->tokenStorage->getToken());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: $logoutEvent then

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 19]
```diff
@@ -90,21 +109,17 @@ public function authenticate(RequestEvent $event)
...
+        $this->eventDispatcher->dispatch($event);
 
-        // handle multiple logout attempts gracefully
-        if ($token = $this->tokenStorage->getToken()) {
-            foreach ($this->handlers as $handler) {
-                $handler->logout($request, $response, $token);
-            }
+        $response = $event->getResponse();
+        if (!$response instanceof Response) {
+            throw new \RuntimeException('No Logout listener set the Response, make sure at least the DefaultLogoutListener is registered.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: logout (lowercase)

File: src/Symfony/Component/Security/Http/Logout/LogoutSuccessHandlerInterface.php [hunk 20]
```diff
@@ -24,6 +24,8 @@
  * LogoutHandlerInterface instead.
  *
  * @author Johannes M. Schmitt <schmittjoh@gmail.com>
+ *
+ * @deprecated since Symfony 5.1.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Logout/LogoutSuccessHandlerInterface.php]: can't we trigger a runtime notice for this one?
should we deprecate DefaultLogoutSuccessHandler?

File: src/Symfony/Component/Security/Http/Tests/Firewall/LogoutListenerTest.php [hunk 21]
```diff
@@ -12,36 +12,43 @@
...
+        });
 
         $httpUtils->expects($this->once())
             ->method('checkRequestPath')
             ->with($request, $options['logout_path'])
             ->willReturn(false);
 
         $listener($event);
+
+        $this->assertFalse($logoutEventDispatched, 'LogoutEvent was dispatched.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Tests/Firewall/LogoutListenerTest.php]: LogoutEvent should not have been dispatched

**Review by nicolas-grekas — COMMENTED** [review]
Here is a superficial review.

File: src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php (hunk 6 above)
Comment (wouterj) [reply src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: Please note that this is for subscribers only, listeners are done before this code.

The current behavior didn't loop over all subscriber tags, only registering an event subscriber once. I figured it was best to keep that behavior. However, I see your point about being able to register on multiple dispatchers, I'll implement it that way 👍

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php (hunk 17 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: IIRC we allowed listening both on named and FQCN events in core, right? Happy to change this to a named event tho.

File: src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php (hunk 8 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php]: Hmm, interesting, seems like we forgot to fix this for the cookie clearing logout handler: https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Security/Http/Logout/CookieClearingLogoutHandler.php

I'll make a separate PR for it.

Comment (wouterj) [reply src/Symfony/Component/Security/Http/EventListener/CookieClearingLogoutListener.php]: See #36252

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php (hunk 17 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: `DeauthenticatedEvent` is not named either. Seems consistent

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 22]
```diff
@@ -90,16 +118,12 @@ public function authenticate(RequestEvent $event)
...
+        $this->eventDispatcher->dispatch($logoutEvent);
 
-        // handle multiple logout attempts gracefully
-        if ($token = $this->tokenStorage->getToken()) {
-            foreach ($this->handlers as $handler) {
-                $handler->logout($request, $response, $token);
-            }
+        $response = $logoutEvent->getResponse();
+        if (!$response instanceof Response) {
+            throw new \RuntimeException('No logout listener set the Response, make sure at least the DefaultLogoutListener is registered.');
```
Comment (linaori) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php]: I feel like the second part of the exception doesn't really belong to a run-time exception: "make sure at least the DefaultLogoutListener is registered." 

What if we reword it to something in form of "Register the DefaultLogoutListener as LogoutEvent listener when no other listener is supposed to give a response"? Probably not the best either, though it does give less of a "you should've registered X" feeling.

**Review by linaori — COMMENTED** [review]
I think this is a really nice change, and I hope that adding this to the event listener documentation will also make this behavior more accessible to developers 😄

File: src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: > Please note that this is for subscribers only, listeners are done before this code.

we should make it work for listeners also, isn't it?

Comment (wouterj) [reply src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: Lines 93 - 98 are for listeners. Listeners have a `method` and `event` attribute, so it already iterates over every tag (see line 67).

Comment (nicolas-grekas) [reply src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: But they don't support a `dispatcher` attribute, do they?

Comment (wouterj) [reply src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: Yes they do :) This is the whole diff, first block adds `dispatcher` for listeners, second block adds `dispatcher` for event subscribers:

Comment (nicolas-grekas) [reply src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: OK got it sorry for the noise :)

Comment (wouterj) [reply src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: No problem!

**Comment by trsteel88** [comment]
+1

I've never understood why the security component uses Handlers instead of events. I'd love to see it using events for everything. For example:

LogoutEvent -> Replaces the need to override the logout handler
AuthenticationFailureEvent -> Replaces the need to override the authentication failure handler
AuthenticationSuccessEvent -> Replaces the need to override the authentication success handler

File: src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php [hunk 23]
```diff
@@ -114,12 +119,27 @@ public function process(ContainerBuilder $container)
...
+            $dispatcherDefinitions = [];
+            foreach ($tags as $attributes) {
+                if (!isset($attributes['dispatcher']) || isset($dispatcherDefinitions[$attributes['dispatcher']])) {
+                    continue;
+                }
+
+                $dispatcherDefinitions[] = $container->getDefinition($attributes['dispatcher']);
+            }
+
+            if ([] === $dispatcherDefinitions) {
```
Comment (fabpot) [inline src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php:131]: `if (!$dispatcherDefinitions) {`

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @wouterj.

File: src/Symfony/Component/Security/Http/Logout/LogoutSuccessHandlerInterface.php [hunk 24]
```diff
@@ -24,6 +27,8 @@
...
```
[the remark this answers is not in the document]
Comment (chalasr) [reply src/Symfony/Component/Security/Http/Logout/LogoutSuccessHandlerInterface.php:31]: fixed in a165ecca73b55f704fda71fe8517d6743884a739

**Comment by stof** [comment]
without seeing your `App\EventListener\LogoutListener`, we cannot help you there.

Also, using `$requestStack->getCurrentRequest()` in the constructor of a service is the wrong way to use the RequestStack. It is totally possible that the constructor is called at the time where the stack is empty (and it will always be the case when using the CLI as there is no processed request there). And the current request will change over time.
`$requestStack->getCurrentRequest()` is meant to be called at the time you need to use the request. Introducing the RequestStack was done precisely to solve problems linked to defining the Request as a service in Symfony 2.x. Getting the request in the constructor reintroduces all these problems (and is even worse, as it does not provide any of the solutions which were implemented in Symfony 2.x for them).

**Comment by Aenos85** [comment]
I agree with you, and in my case I do not need the request. But the class "Symfony\Component\Security\Http\Event\LogoutEvent" is the problem here the constructor is made and it gets the request service. The RequestStack is just a workaround for the listener in the CLI to work at all.

I have already shown my "App\EventListener\LogoutListener" completely, but here again.

And unfortunately the setResponse does not work either.
```
<?php


namespace App\EventListener;

use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\RequestStack;
use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
use Symfony\Component\Security\Http\Event\LogoutEvent;


class LogoutListener extends LogoutEvent
{
    public function __construct(RequestStack $requestStack, ?TokenInterface $token)
    {
        parent::__construct($requestStack->getCurrentRequest(), $token);
    }

    public function onSymfonyComponentSecurityHttpEventLogoutEvent()
    {
        ...
        $this->setResponse(new JsonResponse(["success" => true]));
    }
}
```

The LogoutEvent class have the Problem, withot the following code I would delete me constructor.

```
class LogoutEvent extends Event
{
    private $request;
    private $response;
    private $token;

    public function __construct(Request $request, ?TokenInterface $token)
    {
        $this->request = $request;
        $this->token = $token;
    }
...
```

**Comment by stof** [comment]
LogoutListener should not extend LogoutEvent. the listener and the event are 2 different concepts.

LogoutEvent should be the type of the argument of the method in your listener.

**Comment by wouterj** [comment]
Please read the documentation of the event dispatcher for more information: https://symfony.com/doc/current/event_dispatcher.html

**Comment by Aenos85** [comment]
Okay, I got it. Sorry for the confusion, my bad.

**Comment by ndantonio** [comment]
Hi, I'm new with Symfony. I would like to ask for your help if you can give me an example of using this Logout Event on a listener? What I want to do is if my session expires then it will automatically logout and redirect to login page.

Thank you in advance!

File: src/Symfony/Component/Security/Http/Firewall/LogoutListener.php [hunk 25]
```diff
@@ -90,16 +118,12 @@ public function authenticate(RequestEvent $event)
...
-        if (!$response instanceof Response) {
-            throw new \RuntimeException('Logout Success Handler did not return a Response.');
-        }
+        $logoutEvent = new LogoutEvent($request, $this->tokenStorage->getToken());
+        $this->eventDispatcher->dispatch($logoutEvent);
 
-        // handle multiple logout attempts gracefully
-        if ($token = $this->tokenStorage->getToken()) {
-            foreach ($this->handlers as $handler) {
-                $handler->logout($request, $response, $token);
```
Comment (alexander-schranz) [inline src/Symfony/Component/Security/Http/Firewall/LogoutListener.php:101]: @wouterj 

I'm currently working on the FosHttpCacheBundle which has a [LogoutHandler](https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/396eea2c79661fe591a80788de4d7ae671c76b49/src/Security/Http/Logout/ContextInvalidationLogoutHandler.php#L26) but since Symfony 5.1 this seems not longer be called in the [test](https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/396eea2c79661fe591a80788de4d7ae671c76b49/tests/Functional/Security/Http/Logout/ContextInvalidationLogoutHandlerTest.php#L47):
https://travis-ci.org/github/FriendsOfSymfony/FOSHttpCacheBundle/jobs/714724517

For me it looks like it has todo with this change, but I'm not 100% sure. For the Bundle its not important as that service is deprecated and the thing handled by another listener but maybe it could be important for other project when the LogoutHandler is not called.

Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/LogoutListener.php:101]: That's because it's registered in a [very weird way](https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/9fd07c2256e8974f25a1da2a0c8e616cab3881f3/src/DependencyInjection/FOSHttpCacheExtension.php#L315):

```php
$container->setAlias('security.logout.handler.session', 'fos_http_cache.user_context.session_logout_handler');
```

All internal handlers are now logout listeners, so "overriding" a core service ID no longer works. I'm not sure if internal core service IDs are covered by the BC promise (cc @nicolas-grekas please confirm).

**Comment by TomasVotruba** [comment]
@ndantonio Hi, you can use it like this: https://github.com/wouterj/symfony/blob/a9f096eb1f1fe687596cf62834d6c5efc003ec2a/src/Symfony/Component/Security/Http/EventListener/CsrfTokenClearingLogoutListener.php
